taxjar/taxjar-magento2-extension
# Backup rates delete async

### Context

Users have experienced timeouts due to backup rate sync when a large number of nexus addresses are configured.

### Description

This PR builds on the previously implemented bulk action making rate creation asynchronous.
- Re-uses the existing chunk size already defined for async rate creation.
- Adds constants to the Taxjar Configuration class for message queue topics.
- Refactors many functionalities of the previous rates topic consumer to an abstract class, `AbstractRatesConsumer::class`, which the existing `Consumer::class` now extends from, as well as the newly created `DeleteRatesConsumer::class`
- Configures new topic for bulk action in message queue across various `etc/*.xml`

![Screen Shot 2021-08-04 at 10 31 50 AM](https://user-images.githubusercontent.com/47947793/128215145-c2aeb0ca-e00d-4729-b087-99e22d8cbdc0.png)
_Messages displayed now upon clicking "sync rates"_

### Performance

For some users, the long-running synchronous task of deleting existing backup rates would lead to browser timeout. This change makes the process asynchronous, so no matter the number of backup rates stored, control of the browser will immediately be returned to the admin user while batched deletes process in the background.

### Testing

Prerequisites:
- TaxJar extension enabled in Magento 2
- At least one nexus address configured in TaxJar app and synced in Magento (preferably in a more tax-heavy US region such as California or Colorado)

1. From Admin dashboard, navigate to `Stores>Configuration>Sales>Tax` 
![image](https://user-images.githubusercontent.com/47947793/128218619-1b8569ad-b6de-4aa4-8299-817aa7146fd0.png)
2. In dropdown box for "Backup Rates Enabled" select option "Yes" 
![image](https://user-images.githubusercontent.com/47947793/128218534-01f103c7-43ea-4fa1-a4fb-a4524c27ecd3.png)
3. In top right, click "Save Config". ![image](https://user-images.githubusercontent.com/47947793/128218352-d49fec27-85fd-469d-b819-e48f8f31f327.png)
4. Since there should not be any previous rates, no deletion should occur, but message should be displayed that bulk action to create backup rates has been queued.
5. Refresh the page and observe that backup rates are created
![image](https://user-images.githubusercontent.com/47947793/128218205-6d59a2ad-d553-4cf7-8e3c-2a31bae12c13.png)
6. In the backup rates section, select "Sync Backup Rates" to force-refresh backup rates. 
![image](https://user-images.githubusercontent.com/47947793/128219028-a53f46e7-1faa-4098-9e1e-c4c639d29402.png)
7. Observe bulk action notifications that backup rates are being deleted as well as created. 
![image](https://user-images.githubusercontent.com/47947793/128219276-262cbfa2-923b-4fa9-8043-5f446703cbcb.png)
8. The outcome of delete/create can be verified by inspecting entries on the `tax_calculations`, `tax_calculation_rates` and `tax_calculation_rules` tables, as previous entries should no longer exist.

#### Versions
